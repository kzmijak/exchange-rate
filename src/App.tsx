import "./App.css";
import { Container, Stack, Typography } from "@mui/joy";
import {
  TransactionForm,
  TransactionFormContent,
} from "./modules/TransactionForm";
import { useState } from "react";

function App() {
  const [transactions, setTransactions] = useState<TransactionFormContent[]>(
    []
  );

  const insertTransaction = (transaction: TransactionFormContent) => {
    setTransactions((currTransactions) => [...currTransactions, transaction]);
  };

  return (
    <Container component="body">
      <Stack component="main" spacing={4}>
        <Stack
          component="nav"
          justifyContent="space-between"
          direction="row"
          alignItems="baseline"
          spacing={2}
        >
          <Typography level="h1">List of expenses</Typography>
          <Typography level="h6">1EUR = 4,384 PLN</Typography>
        </Stack>

        <Stack component="section" direction="row">
          <TransactionForm onSubmit={insertTransaction} />
        </Stack>

        {transactions.map((transaction) => (
          <Typography key={transaction.title}>
            {transaction.title} {transaction.amount}
          </Typography>
        ))}
      </Stack>
    </Container>
  );
}

export default App;
