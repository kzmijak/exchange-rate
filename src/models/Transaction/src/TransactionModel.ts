export type TransactionModel = {
  id: string;
  title: string;
  amount: number;
};
