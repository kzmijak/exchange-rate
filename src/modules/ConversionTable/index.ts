export { ConversionTable } from "./src/ConversionTable";
