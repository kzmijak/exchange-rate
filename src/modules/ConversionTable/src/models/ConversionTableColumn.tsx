export const conversionTableColumnConsts = [
  "Title",
  "AmountPLN",
  "AmountEUR",
  "Options",
] as const;

export type ConversionTableColumn =
  (typeof conversionTableColumnConsts)[number];
