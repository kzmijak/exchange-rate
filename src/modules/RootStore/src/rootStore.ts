import { conversionRateStore } from "./stores/ConversionRateStore";

import { transactionsStore } from "./stores/TransactionsStore";

export const createRootStore = () => ({
  conversionRateStore,
  transactionsStore,
});

export type RootStore = ReturnType<typeof createRootStore>;
