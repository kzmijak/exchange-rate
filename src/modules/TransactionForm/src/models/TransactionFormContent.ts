export type TransactionFormContent = {
  title: string;
  amount: number;
};
