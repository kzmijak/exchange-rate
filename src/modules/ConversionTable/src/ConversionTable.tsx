import { Table } from "@mui/joy";
import { FC, useMemo } from "react";
import { TableHeader } from "./components/TableHeader";
import { TableBody } from "./components/TableBody";
import { TransactionModel } from "models/Transaction/src/TransactionModel";
import { arrayMapTransactionModelToConversionTableRow } from "./utils/transactionToRowMapper";
import { TableFooter } from "./components/TableFooter";
import {
  ConversionTableApiContext,
  ConversionTableApiContextType,
} from "./contexts/ConversionTableApiContext";

type ConversionTableProps = {
  transactions: TransactionModel[];
  onTransactionDelete?: (id: TransactionModel["id"]) => void;
};

export const ConversionTable: FC<ConversionTableProps> = ({
  transactions,
  onTransactionDelete = () => null,
}) => {
  const rows = arrayMapTransactionModelToConversionTableRow(transactions);
  const apiContextValue: ConversionTableApiContextType = useMemo(
    () => ({
      deleteRow: onTransactionDelete,
      rows,
    }),
    [transactions]
  );

  return (
    <ConversionTableApiContext.Provider value={apiContextValue}>
      <Table
        aria-label="conversion table"
        borderAxis="none"
        color="neutral"
        size="md"
        stickyHeader
        stripe="odd"
        variant="plain"
        hoverRow
      >
        <TableHeader />
        <TableBody />
        <TableFooter />
      </Table>
    </ConversionTableApiContext.Provider>
  );
};
