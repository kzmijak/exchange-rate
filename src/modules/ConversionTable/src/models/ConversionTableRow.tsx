import { TransactionModel } from "models/Transaction/src/TransactionModel";

export type ConversionTableRow = {
  id: TransactionModel["id"];
  title: TransactionModel["title"];
  amountPLN: TransactionModel["amount"];
};
