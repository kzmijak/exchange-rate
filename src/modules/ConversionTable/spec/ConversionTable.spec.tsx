import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import { ConversionTable } from "../src/ConversionTable";
import { TransactionModel } from "models/TransactionModel";

const transactions = [{}, {}, {}] as TransactionModel[];

describe("ConversionTable", () => {
  const Component = () => <ConversionTable transactions={transactions} />;

  it("should match snapshot", () => {
    expect(render(<Component />).container).toMatchSnapshot();
  });
});
