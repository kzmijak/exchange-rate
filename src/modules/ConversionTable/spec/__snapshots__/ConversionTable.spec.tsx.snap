// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConversionTable > should match snapshot 1`] = `
<div>
  <table
    aria-label="conversion table"
    class="MuiTable-root MuiTable-stickyHeader MuiTable-stickyFooter MuiTable-hoverRow MuiTable-borderAxisNone MuiTable-variantPlain MuiTable-colorNeutral MuiTable-sizeMd css-1urnli7-JoyTable-root"
  >
    <thead>
      <tr>
        <th
          class="MuiBox-root css-34k09d"
        >
          <p
            class="MuiTypography-root MuiTypography-inherit css-6zqddh-JoyTypography-root"
          >
            Title
          </p>
        </th>
        <th
          class="MuiBox-root css-yikrym"
        >
          <p
            class="MuiTypography-root MuiTypography-inherit css-1sod2m5-JoyTypography-root"
          >
            Amount (PLN)
          </p>
        </th>
        <th
          class="MuiBox-root css-yikrym"
        >
          <p
            class="MuiTypography-root MuiTypography-inherit css-1sod2m5-JoyTypography-root"
          >
            Amount (EUR)
          </p>
        </th>
        <th
          class="MuiBox-root css-yikrym"
        >
          <p
            class="MuiTypography-root MuiTypography-inherit css-12jjlxs-JoyTypography-root"
          >
            Options
          </p>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <p
            class="MuiTypography-root MuiTypography-inherit css-6zqddh-JoyTypography-root"
          />
        </td>
        <td>
          <p
            class="MuiTypography-root MuiTypography-inherit css-1sod2m5-JoyTypography-root"
          />
        </td>
        <td>
          <p
            class="MuiTypography-root MuiTypography-inherit css-1sod2m5-JoyTypography-root"
          >
            NaN
          </p>
        </td>
        <td
          align="center"
        >
          <button
            class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeSm css-1gh95q6-JoyButton-root"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <p
            class="MuiTypography-root MuiTypography-inherit css-6zqddh-JoyTypography-root"
          />
        </td>
        <td>
          <p
            class="MuiTypography-root MuiTypography-inherit css-1sod2m5-JoyTypography-root"
          />
        </td>
        <td>
          <p
            class="MuiTypography-root MuiTypography-inherit css-1sod2m5-JoyTypography-root"
          >
            NaN
          </p>
        </td>
        <td
          align="center"
        >
          <button
            class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeSm css-1gh95q6-JoyButton-root"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
      <tr>
        <td>
          <p
            class="MuiTypography-root MuiTypography-inherit css-6zqddh-JoyTypography-root"
          />
        </td>
        <td>
          <p
            class="MuiTypography-root MuiTypography-inherit css-1sod2m5-JoyTypography-root"
          />
        </td>
        <td>
          <p
            class="MuiTypography-root MuiTypography-inherit css-1sod2m5-JoyTypography-root"
          >
            NaN
          </p>
        </td>
        <td
          align="center"
        >
          <button
            class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeSm css-1gh95q6-JoyButton-root"
            type="button"
          >
            Delete
          </button>
        </td>
      </tr>
    </tbody>
    <tfoot>
      <tr>
        <td
          colspan="4"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1wlnfom-JoyTypography-root"
          >
            Sum: 
            NaN
             PLN ( 
            NaN
             EUR )
          </p>
        </td>
      </tr>
    </tfoot>
  </table>
</div>
`;
